### PR TITLE
Tpwdcee 5860 php-enum

### PR DESCRIPTION
**Pull request by flavius**
After this change is merged in, https://github.com/wirecard/magento2-ee/pull/356 should also be released.

Also change the base branch.

### CODE REVIEW

File: src/Helper/Enum.php [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+     *
+     * @return mixed
+     * @link http://php.net/manual/en/jsonserializable.jsonserialize.php
+     */
+    public function jsonSerialize()
+    {
+        return $this->getValue();
+    }
+
+}
```
Comment (tomazpu) [inline src/Helper/Enum.php]: Codacy Issue found: [Expected 1 newline at end of file; 0 found](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173)

**Comment by tomazpu** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- src/Helper/Enum.php  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173&bid=16312560)

File: src/Helper/Enum.php (hunk 1 above)
Comment (tomazpu) [inline src/Helper/Enum.php]: Codacy Issue found: [The closing brace for the class must go on the next line after the body](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173)

File: src/Helper/Enum.php [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+/**
+ * Base Enum class
+ *
+ * Create an enum by implementing this class and adding class constants.
+ *
+ * @author Matthieu Napoli <matthieu@mnapoli.fr>
+ * @author Daniel Costa <danielcosta@gmail.com>
+ * @author Mirosław Filip <mirfilip@gmail.com>
+ */
+abstract class Enum
```
Comment (tomazpu) [inline src/Helper/Enum.php]: Codacy Issue found: [The class Enum has 11 non-getter- and setter-methods. Consider refactoring Enum to keep number of methods under 10.](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173)

**Comment by tomazpu** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/Helper/Enum.php  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173&bid=16312560)

File: src/Helper/Enum.php (hunk 2 above)
Comment (tomazpu) [inline src/Helper/Enum.php]: Codacy Issue found: [The class Enum has 11 non-getter- and setter-methods. Consider refactoring Enum to keep number of methods under 10.](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173)

File: src/Helper/Enum.php [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+     * natively.
+     *
+     * @return mixed
+     * @link http://php.net/manual/en/jsonserializable.jsonserialize.php
+     */
+    public function jsonSerialize()
+    {
+        return $this->getValue();
+    }
+}
```
Comment (tomazpu) [inline src/Helper/Enum.php]: Codacy Issue found: [Expected 1 newline at end of file; 0 found](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173)

**Comment by tomazpu** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/Helper/Enum.php  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173&bid=16312560)

File: src/Helper/Enum.php [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
```
Comment (tomazpu) [inline src/Helper/Enum.php]: Codacy Issue found: [Expected 1 newline at end of file; 0 found](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173)

File: src/Helper/Enum.php [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+ * Base Enum class
+ *
+ * Create an enum by implementing this class and adding class constants.
+ *
+ * @author Matthieu Napoli <matthieu@mnapoli.fr>
+ * @author Daniel Costa <danielcosta@gmail.com>
+ * @author Mirosław Filip <mirfilip@gmail.com>
+ * @SuppressWarnings(PHPMD.ExcessivePublicCount)
+ */
+abstract class Enum
```
Comment (tomazpu) [inline src/Helper/Enum.php]: Codacy Issue found: [The class Enum has 11 non-getter- and setter-methods. Consider refactoring Enum to keep number of methods under 10.](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173)

**Comment by tomazpu** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/Helper/Enum.php  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173&bid=16312560)

File: src/Helper/Enum.php [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+ * Base Enum class
+ *
+ * Create an enum by implementing this class and adding class constants.
+ *
+ * @SuppressWarnings(PHPMD.ExcessivePublicCount)
+ * @author Matthieu Napoli <matthieu@mnapoli.fr>
+ * @author Daniel Costa <danielcosta@gmail.com>
+ * @author Mirosław Filip <mirfilip@gmail.com>
+ */
+abstract class Enum
```
Comment (tomazpu) [inline src/Helper/Enum.php]: Codacy Issue found: [The class Enum has 11 non-getter- and setter-methods. Consider refactoring Enum to keep number of methods under 10.](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173)

File: src/Helper/Enum.php [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+     * natively.
+     *
+     * @return mixed
+     * @link http://php.net/manual/en/jsonserializable.jsonserialize.php
+     */
+    public function jsonSerialize()
+    {
+        return $this->getValue();
+    }
+}
```
Comment (tomazpu) [inline src/Helper/Enum.php]: Codacy Issue found: [Expected 1 newline at end of file; 0 found](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173)

**Comment by tomazpu** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/Helper/Enum.php  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173&bid=16312560)

File: src/Helper/Enum.php [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
```
Comment (tomazpu) [inline src/Helper/Enum.php]: Codacy Issue found: [Expected 1 newline at end of file; 0 found](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173)

**Comment by tomazpu** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ src/BackendService.php  6
         

Complexity increasing per file
==============================
- src/Helper/Enum.php  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173&bid=16312560)

File: src/Helper/Enum.php [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    /**
+     * Compares one Enum with another.
+     *
+     * This method is final, for more information read https://github.com/myclabs/php-enum/issues/4
+     *
+     * @return bool True if Enums are equal, false if not equal
+     */
+    final public function equals(\MyCLabs\Enum\Enum $enum = null)
```
[the remark this answers is not in the document]
Comment (flavius) [reply src/Helper/Enum.php]: Good catch!

**Comment by tomazpu** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ src/BackendService.php  6
         

Complexity increasing per file
==============================
- src/Helper/Enum.php  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173&bid=16312560)

**Comment by tomazpu** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ src/Response/XmlResponseData.php  1
+ src/Entity/TransactionDetails.php  1
+ src/Entity/CustomFieldCollection.php  1
+ src/Entity/CustomField.php  2
+ src/Entity/RiskInfo.php  2
+ src/Transaction/Transaction.php  1
         

Coverage decreased per file
===========================
- src/Entity/Browser.php  -2
- src/Response/ResponseFactory.php  -9
- src/Entity/Basket.php  -2
- src/Entity/Payload/NvpPayloadData.php  -8
- src/Transaction/CustomFieldTransaction.php  -15
- src/Entity/Redirect.php  -8
- src/TransactionService.php  -2
- src/Mapper/RequestMapper.php  -4
- src/Helper/XmlBuilder.php  -10
- src/Entity/Mappable.php  -1
- src/Entity/Device.php  -4
- src/Entity/AccountHolder.php  -1
- src/Transaction/EpsTransaction.php  -10
- src/Config/Config.php  -2
         

Complexity increasing per file
==============================
- src/Helper/Enum.php  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173&bid=16312560)

**Comment by tomazpu** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ src/BackendService.php  21
+ src/Transaction/CustomFieldTransaction.php  10
+ src/Response/ResponseFactory.php  4
         

Complexity increasing per file
==============================
- src/Helper/Enum.php  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/paymentSDK-php/pullRequest?prid=4932173&bid=16312560)

**Pull request merged**